src/riak_kv_qry_buffers.erl

```diff
@@ -420,55 +427,97 @@ do_batch_put(QBufRef, Data, #state{qbufs          = QBufs0,
+                total_records = TotalRecords0,
+                key_field_positions = KeyFieldPositions} = QBuf0,
+          Data,
+          TotalSize, HardWatermark, InmemMax) ->
     ChunkSize = compute_chunk_size(Data),
     case TotalSize + ChunkSize > HardWatermark of
         true ->
             {error, total_qbuf_size_limit_reached};
         false ->
+            %% update stats/counters
+            ChunkId = ChunksGot0,
+            ChunksGot = ChunksGot0 + 1,
             %% ChunkId will be used to construct a new and unique key
             %% for each record. Ideally, this should be the serial
             %% number of the subquery.
-            ChunkId = ChunksGot0,
-            ChunksGot = ChunksGot0 + 1,
-            lager:debug("adding chunk ~b of ~b", [ChunksGot, ChunksNeed]),
+            IsReady = (ChunksNeed == ChunksGot),
+
+            %% construct keys for new data
             OrdByFieldQualifiers = lists:map(fun get_ordby_field_qualifiers/1, OrderBy),
-            case riak_kv_qry_buffers_ldb:add_rows(LdbRef, Data, ChunkId,
-                                                  KeyFieldPositions,
-                                                  OrdByFieldQualifiers) of
-                ok ->
-                    IsReady = (ChunksNeed == ChunksGot),
-                    QBuf = QBuf0#qbuf{size          = Size + ChunkSize,
-                                      chunks_got    = ChunksGot,
-                                      total_records = TotalRecords0 + length(Data),
-                                      is_ready      = IsReady,
-                                      last_accessed = os:timestamp()},
+            KeyedData = enkey(Data, KeyFieldPositions, OrdByFieldQualifiers, ChunkId),
+
+            QBuf1 = QBuf0#qbuf{size          = Size + ChunkSize,
+                               chunks_got    = ChunksGot,
+                               total_records = TotalRecords0 + length(Data),
+                               is_ready      = IsReady,
+                               last_accessed = os:timestamp()},
+
+            HaveLdbOpened = not is_function(LdbRefOrDelayedCreateFun),
+            InmemAcc = lists:append(InmemBuffer0, KeyedData),
+            EleveldbPutF =
+                fun(LdbRef) ->
+                        lager:debug("adding chunk ~b of ~b to ldb", [ChunksGot, ChunksNeed]),
+                        case riak_kv_qry_buffers_ldb:add_rows(LdbRef, InmemAcc) of
+                            ok ->
+                                QBuf = QBuf1#qbuf{inmem_buffer = [],
+                                                  ldb_ref      = LdbRef},
+                                {ok, QBuf};
+                            {error, _} = ErrorReason ->
+                                ErrorReason
+                        end
+                end,
+
+            case (HaveLdbOpened orelse not can_afford_inmem(InmemMax)) of
+                false ->
+                    InmemBuffer =
```

Comment:
There is no regard for crash recovery for query buffers by design.

Query buffer manager runs on the same node where coordinator runs; if it goes down, the client will lose the query, and when the same query is reissued, the recovered buffer will be of no use because all query ids are unique (they are constructed with `term_to_binary(make_ref())`).

If the riak_kv_qry_buffers application were to crash on its own, it will purge any existing query buffers on subsequent init.

And if an IO error occurs, sending the client an "internal qbuf error" is the lesser evil than blocking the query buffer manager gen_server while recovery is in progress.

Whether ETS can lessen the burden on gc is an open question, but any gains here would have to be evaluated against (1) more code to execute (`lists:sublist(Acc, Offset, Limit)` vs lots of `ets:lookup`s or a `:fold` with counter), (2) increased code complexity.